hypothesis-python/src/hypothesis/_strategies.py

```diff
@@ -525,11 +528,19 @@ def floats(
         exclude_min or (min_arg is not None and min_value < min_arg)
     ):
         min_value = next_up(min_value, width)
-        assert min_value > min_arg or min_value == min_arg == 0  # type: ignore
+        if min_value == min_arg:
+            assert min_value == min_arg == 0
+            assert is_negative(min_arg) and not is_negative(min_value)
+            min_value = next_up(min_value, width)
+        assert min_value > min_arg  # type: ignore
     if max_value is not None and (
         exclude_max or (max_arg is not None and max_value > max_arg)
     ):
         max_value = next_down(max_value, width)
+        if max_value == max_arg:
+            assert max_value == max_arg == 0
+            assert is_negative(max_value) and not is_negative(max_arg)
+            max_value = next_down(max_value, width)
         assert max_value < max_arg or max_value == max_arg == 0  # type: ignore
```

Comment:
By this point in the code, there should be a stricter assertion:
```suggestion
        assert max_value < max_arg  # type: ignore
```